tests/5.2/target/test_target_map_device_pointer.c

```diff
@@ -0,0 +1,64 @@
+//--------------- test_target_map_device_pointer.c ---------------------------//
+// OpenMP API Version 5.2 Nov 2021
+// Pg. 627, line 24
+// This test ensures that--paraphrase of Section 5.8.6--a pointer that is
+// implicitly mapped or explicitely mapped to the device, predetermined to be
+// firstprivate in the device data environment, will retain its original value
+// (attained prior to the target region) and follow the semantics of the
+// firstprivate clause. This is true if a matching mapped list item, as defined
+// in Section 5.8.6, is not found for the pointer. In the test, the pointer p1
+// is initialized with the base array address of arr2. When p1 is implicitly
+// mapped to the device, it should retain the same base array address, as the
+// array itself is not mapped (a matching mapped list item is not found). This
+// check is made in the target region with a mapped integer variable holding the
+// base address of arr2. The pointer p2 is initialized to NULL, and is checked
+// to remain NULL after being initialized on a device. Following the checks, p1
+// is changed to NULL on the device and checked outside of the target region if
+// it was not changed from its initial value of the base address of arr2 (per
+// firstprivate semnatics). p2 is given an arbitrary value for the purpose of
+// ensuring that it too retains its original value after the target region. A
+// defaultmap clause is present on the target construct to explicitely ensure
+// that the default semantics of mapping apply to all mapped (implicit/explicit)
+// pointer types. The zero length array, arr1, is mapped to the device to ensure
+// that a runtime check occurs in accordance with the semantics described in the
+// target construct on page 285 of the specifications.
+//----------------------------------------------------------------------------//
+#include "ompvv.h"
+#include <omp.h>
+#include <stdint.h> //includes intptr_t
+
+#define N 16
+
+int test_target_pointer() {
+  int errors = 0;
+  int arr1[0], arr2[N];
```

Comment:
Why `int arr1[0]`?

Reply:
omp examples 5.2: 6.3 Pointer Mapping
21 This scenario can be further split into two cases: the list item is a
22 zero-length array section (e.g., p[:0]) or it is not.
23 If it is a zero-length array section, this will trigger a runtime check on entry to the target region
24 for a previously mapped list item where the value of the pointer falls within the range of its base
25 address and ending address. If such a match is found the private pointer is initialized to the device
26 address corresponding to the value of the original pointer, and otherwise it is initialized to NULL
27 (or retains its original value if the unified_address requirement is specified for that
28 compilation unit).
29 If the list item (again, call it the pointee) is not a zero-length array section, the private pointer will
30 be initialized such that references in the target region to the pointee list item that use the pointer
31 will access the corresponding pointee.
Note^: Pointer is not initialized to NULL as stated (openmp forgot to update this part)

Reply:
It seems that the zero-length array is a GCC extension but not a part of the C standard (https://gcc.gnu.org/onlinedocs/gcc/Zero-Length.html), which also says the following:

> Declaring zero-length arrays in other contexts, including as interior members of structure objects or as non-member objects, is discouraged. Accessing elements of zero-length arrays declared in such contexts is undefined and may be diagnosed.